packages/shared/edge-utils/src/inject.ts

```diff
@@ -0,0 +1,232 @@
+  const uuidKey = context.config.uuidKey;
+  const uuid = attributes[uuidKey];
+  const trackingCallback = context.config.growthbook.trackingCallback;
+  const blockedChangeIds = getBlockedExperiments({
+    context,
+    experiments,
+    completedChangeIds,
+    preRedirectChangeIds,
+  });
+  const injectRedirectUrlScript = context.config.injectRedirectUrlScript;
+  const enableStreaming = context.config.enableStreaming;
+  const enableStickyBucketing = context.config.enableStickyBucketing;
+  const stickyAssignments = growthbook.getStickyBucketAssignmentDocs();
+
+  const gbContext: Omit<GbContext, "trackingCallback"> & {
+    uuidCookieName: string;
+    uuidKey: string;
+    uuid?: string;
+    attributeKeys?: Record<string, string>;
+    persistUuidOnLoad?: boolean;
+    useStickyBucketService?: "cookie" | "localStorage";
+    stickyBucketPrefix?: string;
+    trackingCallback: string; // replaced by macro
+    payload?: FeatureApiResponse;
+    noStreaming?: boolean;
+  } = {
+    uuidCookieName,
+    uuidKey,
+    uuid,
+    persistUuidOnLoad: true,
+    attributes,
+    trackingCallback: "__TRACKING_CALLBACK__",
+    payload: sdkPayload,
+    disableVisualExperiments: ["skip", "edge"].includes(
+      context.config.runVisualEditorExperiments,
+    ),
+    disableJsInjection: context.config.disableJsInjection,
+    disableUrlRedirectExperiments: ["skip", "edge"].includes(
+      context.config.runUrlRedirectExperiments,
+    ),
+    disableCrossOriginUrlRedirectExperiments: ["skip", "edge"].includes(
+      context.config.runCrossOriginUrlRedirectExperiments,
+    ),
+    jsInjectionNonce: nonce,
+    blockedChangeIds,
+    noStreaming: !enableStreaming,
+    useStickyBucketService: enableStickyBucketing ? "cookie" : undefined,
+    stickyBucketPrefix: context.config.stickyBucketPrefix,
+    stickyBucketAssignmentDocs: stickyAssignments,
+  };
+
+  let scriptTag = `
+<script
+  data-api-host="${context.config.growthbook.apiHost}"
+  data-client-key="${context.config.growthbook.clientKey}"${
+    context.config.growthbook.decryptionKey
+      ? `\n  data-decryption-key="${context.config.growthbook.decryptionKey}"`
+      : ""
+  }${nonce ? `\n  nonce="${nonce}"` : ""}
+>
+  window.growthbook_config = ${JSON.stringify(gbContext)};
+${
+  deferredTrackingCalls?.length
+    ? `
+  window.growthbook_queue = [
+    (gb) => {
+      gb.setDeferredTrackingCalls(${JSON.stringify(
+        scrubInvalidTrackingCalls(deferredTrackingCalls, preRedirectChangeIds),
+      )});
+      gb.fireDeferredTrackingCalls();
+    }
+  ];
+`
+    : ""
+}
+  ${sdkWrapper}
+  ${
+    url !== oldUrl && injectRedirectUrlScript
+      ? `window.history.replaceState(undefined, undefined, ${JSON.stringify(
+          url,
+        )})`
+      : ""
+  }
+</script>
+`;
+  scriptTag = scriptTag.replace(
+    `"__TRACKING_CALLBACK__"`,
+    trackingCallback || "undefined",
+  );
+
+  const pattern = context.config.scriptInjectionPattern || "</body>";
```

Comment:
I think we want to inject before the closing `</head>` tag instead by default, right?  Otherwise, there will be a flicker if any tests are being run.

Reply:
good point, fixed